File: bot-intelligence/src/main/java/gg/mineral/bot/ai/goal/EatGappleGoal.java
package gg.mineral.bot.ai.goal;

import gg.mineral.bot.api.controls.Key;
import gg.mineral.bot.api.controls.MouseButton;
import gg.mineral.bot.api.entity.effect.PotionEffectType;
import gg.mineral.bot.api.entity.living.player.ClientPlayer;

import gg.mineral.bot.api.event.Event;
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent;
import gg.mineral.bot.api.goal.Goal;

import gg.mineral.bot.api.instance.ClientInstance;
import gg.mineral.bot.api.inv.item.Item;

import gg.mineral.bot.api.screen.type.ContainerScreen;

import lombok.Getter;
import lombok.val;

@Getter
public class EatGappleGoal extends Goal {

    private boolean inventoryOpen = false, eating = false;

    @Override
    public boolean shouldExecute() {
        var hasRegen = false;
        val regenId = PotionEffectType.REGENERATION.getId();
        val fakePlayer = clientInstance.getFakePlayer();
        val activeIds = fakePlayer.getActivePotionEffectIds();

        for (int i = 0; i < activeIds.length; i++)
            if (activeIds[i] == regenId)
                hasRegen = true;

        boolean shouldExecute = eating || canSeeEnemy() && hasGapple() && !hasRegen;
        info(this, "Checking shouldExecute: " + shouldExecute);
        return shouldExecute;
    }

    public EatGappleGoal(ClientInstance clientInstance) {
        super(clientInstance);
        info(this, "EatGappleGoal initialized");
    }

    private boolean hasGapple() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null");
            return false;
        }

        boolean hasGapple = inventory.contains(Item.GOLDEN_APPLE);
        info(this, "Has golden apple: " + hasGapple);
        return hasGapple;
    }

    private boolean canSeeEnemy() {
        val fakePlayer = clientInstance.getFakePlayer();
        val world = fakePlayer.getWorld();

        if (world == null) {
            warn(this, "World is null");
            return false;
        }

        boolean canSeeEnemy = world.getEntities().stream()
                .anyMatch(entity -> entity instanceof ClientPlayer
                        && !clientInstance.getConfiguration().getFriendlyUUIDs().contains(entity.getUuid()));
        info(this, "Checking canSeeEnemy: " + canSeeEnemy);
        return canSeeEnemy;
    }

    private void eatGapple() {
        this.eating = true;
        success(this, "Started eating golden apple");
    }

    private void switchToGapple() {
        eating = false;
        info(this, "Switching to golden apple");
        var gappleSlot = -1;
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null");
            return;
        }

        for (int i = 0; i < 36; i++) {
            val itemStack = inventory.getItemStackAt(i);
            if (itemStack == null)
                continue;
            val item = itemStack.getItem();
            if (item.getId() == Item.GOLDEN_APPLE) {
                gappleSlot = i;
                break;
            }
        }

        if (gappleSlot > 8) {
            if (!inventoryOpen) {
                inventoryOpen = true;
                pressKey(10, Key.Type.KEY_E);
                info(this, "Opened inventory to access golden apple");
                return;
            }

            val screen = clientInstance.getCurrentScreen();

            if (screen == null) {
                inventoryOpen = false;
                warn(this, "Screen is null; closing inventory");
                return;
            }

            val inventoryContainer = fakePlayer.getInventoryContainer();

            if (inventoryContainer == null) {
                inventoryOpen = false;
                warn(this, "Inventory container is null; closing inventory");
                return;
            }

            val slot = inventoryContainer.getSlot(inventory, gappleSlot);

            if (slot == null) {
                inventoryOpen = false;
                pressKey(10, Key.Type.KEY_ESCAPE);
                warn(this, "Golden apple slot is null; closing inventory");
                return;
            }

            if (screen instanceof ContainerScreen containerScreen) {
                val x = containerScreen.getSlotX(slot);
                val y = containerScreen.getSlotY(slot);

                val currX = getMouseX();
                val currY = getMouseY();

                if (currX != x || currY != y) {
                    setMouseX(x);
                    setMouseY(y);
                    info(this, "Moving mouse to golden apple slot at x: " + x + ", y: " + y);
                } else {
                    pressKey(10, Key.Type.KEY_8);
                    success(this, "Moved to end slot for golden apple");
                }
            }
            return;
        }

        if (inventoryOpen) {
            inventoryOpen = false;
            pressKey(10, Key.Type.KEY_ESCAPE);
            info(this, "Closing inventory after switching to golden apple");
            return;
        }

        pressKey(10, Key.Type.valueOf("KEY_" + (gappleSlot + 1)));
        success(this, "Switched to golden apple slot: " + (gappleSlot + 1));
    }

    @Override
    public void onTick() {
        val fakePlayer = clientInstance.getFakePlayer();
        val inventory = fakePlayer.getInventory();

        if (inventory == null) {
            warn(this, "Inventory is null on tick");
            return;
        }

        var hasRegen = false;
        val regenId = PotionEffectType.REGENERATION.getId();
        val activeIds = fakePlayer.getActivePotionEffectIds();

        for (int i = 0; i < activeIds.length; i++)
            if (activeIds[i] == regenId) {
                hasRegen = true;
                break;
            }

        if (eating && hasRegen) {
            eating = false;
            info(this, "Stopped eating as regeneration is active");
        }

        val rmbHeld = getButton(MouseButton.Type.RIGHT_CLICK).isPressed();

        if (!eating && rmbHeld) {
            unpressButton(MouseButton.Type.RIGHT_CLICK);
            info(this, "Unpressed RIGHT_CLICK as eating stopped");
        }

        if (hasRegen)
            return;

        if (eating && !rmbHeld) {
            pressButton(MouseButton.Type.RIGHT_CLICK);
            info(this, "Pressed RIGHT_CLICK for eating golden apple");
        }

        if (!delayedTasks.isEmpty())
            return;

        val itemStack = inventory.getHeldItemStack();

        if (itemStack != null && itemStack.getItem().getId() == Item.GOLDEN_APPLE) {
            schedule(() -> eatGapple(), 100);
            info(this, "Scheduled eatGapple task");
        } else {
            schedule(() -> switchToGapple(), 100);
            info(this, "Scheduled switchToGapple task");
        }
    }

    @Override
    public boolean onEvent(Event event) {
        if (event instanceof MouseButtonEvent mouseButtonEvent) {
            if (eating && mouseButtonEvent.getType() == MouseButton.Type.RIGHT_CLICK && !mouseButtonEvent.isPressed()) {
                info(this, "Ignoring RIGHT_CLICK release event while eating");
                return true;
            }
        }
        return false;
    }

    @Override
    public void onGameLoop() {
    }
}

File: bot-intelligence/src/main/java/gg/mineral/bot/ai/goal/EatGappleGoal.kt
package gg.mineral.bot.ai.goal

import gg.mineral.bot.ai.goal.type.InventoryGoal
import gg.mineral.bot.api.controls.Key
import gg.mineral.bot.api.controls.MouseButton
import gg.mineral.bot.api.entity.effect.PotionEffectType
import gg.mineral.bot.api.entity.living.ClientLivingEntity
import gg.mineral.bot.api.entity.living.player.ClientPlayer
import gg.mineral.bot.api.event.Event
import gg.mineral.bot.api.event.peripherals.MouseButtonEvent
import gg.mineral.bot.api.goal.Sporadic
import gg.mineral.bot.api.goal.Timebound
import gg.mineral.bot.api.instance.ClientInstance
import gg.mineral.bot.api.inv.item.Item

class EatGappleGoal(clientInstance: ClientInstance) : InventoryGoal(clientInstance), Sporadic, Timebound {
    override var executing: Boolean = false
    override var startTime: Long = 0
    override val maxDuration: Long = 100
    private var eating = false

    override fun shouldExecute(): Boolean {
        var hasRegen = false
        val regenId = PotionEffectType.REGENERATION.id
        val fakePlayer = clientInstance.fakePlayer
        val activeIds = fakePlayer.activePotionEffectIds

        for (activeId in activeIds) if (activeId == regenId) {
            hasRegen = true
            break
        }

        val shouldExecute =
            canSeeEnemy() && hasGapple() && !hasRegen && (fakePlayer.health < 10 || distanceAwayFromEnemies() in 8.0..16.0)
        logger.debug("Checking shouldExecute: $shouldExecute")
        return shouldExecute
    }

    override fun onStart() {
        pressKey(Key.Type.KEY_W, Key.Type.KEY_LCONTROL)
        unpressKey(Key.Type.KEY_S, Key.Type.KEY_A, Key.Type.KEY_D)
    }

    init {
        logger.debug("EatGappleGoal initialized")
    }

    private fun hasGapple(): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        val hasGapple = inventory.contains(Item.GOLDEN_APPLE)
        logger.debug("Has golden apple: $hasGapple")
        return hasGapple
    }

    private fun canSeeEnemy(): Boolean {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        val canSeeEnemy = world.entities.any {
            it is ClientPlayer
                    && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid)
        }
        logger.debug("Checking canSeeEnemy: $canSeeEnemy")
        return canSeeEnemy
    }

    private fun angleAwayFromEnemies(): Float {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        val enemy = world.entities
            .minByOrNull {
                if (it is ClientLivingEntity && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid))
                    it.distance3DTo(fakePlayer)
                else Double.MAX_VALUE
            } ?: return fakePlayer.yaw
        val x: Double = enemy.x - fakePlayer.x
        val z: Double = enemy.z - fakePlayer.z

        var yaw = Math.toDegrees(-fastArcTan(x / z)).toFloat()
        if (z < 0.0 && x < 0.0) yaw = (90.0 + Math.toDegrees(fastArcTan(z / x))).toFloat()
        else if (z < 0.0 && x > 0.0) yaw = (-90.0 + Math.toDegrees(fastArcTan(z / x))).toFloat()
        return yaw + 180.0f
    }

    private fun distanceAwayFromEnemies(): Double {
        val fakePlayer = clientInstance.fakePlayer
        val world = fakePlayer.world

        return world.entities
            .minOfOrNull {
                if (it is ClientLivingEntity && !clientInstance.configuration.friendlyUUIDs.contains(it.uuid))
                    it.distance3DTo(fakePlayer)
                else Double.MAX_VALUE
            } ?: Double.MAX_VALUE
    }

    private fun getGappleSlot(): Int {
        var gappleSlot = -1
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        for (i in 0..35) {
            val itemStack = inventory.getItemStackAt(i) ?: continue
            val item = itemStack.item
            if (item.id == Item.GOLDEN_APPLE) {
                gappleSlot = i
                break
            }
        }

        return gappleSlot
    }

    override fun onTick(tick: Tick) {
        val gappleSlot = getGappleSlot()
        val fakePlayer = clientInstance.fakePlayer
        val inventory = fakePlayer.inventory

        tick.finishIf("Valid gapple slot not found", gappleSlot == -1)

        tick.prerequisite("In Hotbar", gappleSlot <= 8) {
            moveItemToHotbar(gappleSlot, inventory)
        }

        tick.prerequisite("Inventory Closed", !inventoryOpen) {
            inventoryOpen = false
        }

        tick.prerequisite("Correct Hotbar Slot Selected", inventory.heldSlot == gappleSlot) {
            pressKey(10, Key.Type.valueOf("KEY_" + (gappleSlot + 1)))
        }

        tick.finishIf("Not Holding Valid Gapple", inventory.heldItemStack?.item?.id != Item.GOLDEN_APPLE)

        tick.finishIf("Has Regen", fakePlayer.activePotionEffectIds.any { it == PotionEffectType.REGENERATION.id })

        tick.prerequisite("Eating", eating && getButton(MouseButton.Type.RIGHT_CLICK).isPressed) {
            pressButton(MouseButton.Type.RIGHT_CLICK)
            eating = true
        }

        tick.execute {
            if (distanceAwayFromEnemies() < 16) {
                setMouseYaw(angleAwayFromEnemies())
                pressKey(Key.Type.KEY_SPACE, Key.Type.KEY_W, Key.Type.KEY_LCONTROL)
            }
        }
    }

    override fun onEnd() {
        eating = false
        unpressButton(MouseButton.Type.RIGHT_CLICK)
        unpressKey(Key.Type.KEY_SPACE)
        if (inventoryOpen) {
            inventoryOpen = false
            logger.debug("Closing inventory after eating golden apple")
        }
    }

    override fun onEvent(event: Event): Boolean {
        if (event is MouseButtonEvent) {
            if (eating && event.type == MouseButton.Type.RIGHT_CLICK && !event.pressed) {
                logger.debug("Ignoring RIGHT_CLICK release event while eating")
                return true
            }
        }
        return false
    }

    public override fun onGameLoop() {
    }
}
